### PR TITLE
Persist namespaces

### PR DESCRIPTION
**Pull request by Snaipe**
Mostly creating this so that I have an easier time reviewing/commenting

### CODE REVIEW

File: enter.c [hunk 1]
```diff
@@ -36,68 +37,59 @@
...
-	};
+	return flags[ns].proc_ns_name;
+}
+	
+
+struct nsaction {
+	/* Which namespace is being unshared or entered */
+	int flag;
+
+	/* If enter_fd is -1, then the namespace given by "flag" is to
```
Comment (Snaipe) [inline enter.c:65]: No longer accurate, maybe say that it can be >=0 or either NSACTION_*?

File: enter.c [hunk 2]
```diff
@@ -110,24 +102,51 @@ int enter(struct entry_settings *opts)
...
+	
+	int nsactions[MAX_SHARES];
+	opts_to_nsactions(opts, nsactions);
+
+	struct outer_helper outer_helper;
+	outer_helper.persist = opts->persist;
+	outer_helper.unshare_user = nsactions[SHARE_USER] == NSACTION_UNSHARE;
+	outer_helper_spawn(&outer_helper);
+
+	/* While we still have privileges, setns into all relevant
```
Comment (Snaipe) [inline enter.c:127]: This really should be done unprivileged. Did you have time to look at my PoC?

File: enter.h [hunk 3]
```diff
@@ -19,12 +19,31 @@ enum {
 	/* this is a very generous upper bound for the number of supported
```
Comment (Snaipe) [inline enter.h:19]: Remove that comment

File: main.c [hunk 4]
```diff
@@ -67,26 +104,38 @@ int main(int argc, char *argv[], char *envp[])
...
-		{ "uid",        required_argument,  NULL,           OPTION_UID      },
-		{ "gid",        required_argument,  NULL,           OPTION_GID      },
-		{ "groups",     required_argument,  NULL,           OPTION_GROUPS   },
-		{ "arch",       required_argument,  NULL,           OPTION_ARCH     },
-		{ "share",      required_argument,  NULL,           OPTION_SHARE    },
-		{ "argv0",      required_argument,  NULL,           OPTION_ARGV0    },
-		{ "hostname",   required_argument,  NULL,           OPTION_HOSTNAME },
-		{ "domainname", required_argument,  NULL,           OPTION_DOMAIN   },
-		{ "time",       required_argument,  NULL,           OPTION_TIME     },
+		{ "workdir",      required_argument, NULL, OPTION_WORKDIR      },
```
Comment (Snaipe) [inline main.c:107]: Micro-nit: I typically align the required/optional/no_argument field on a 8-char width boundary, that leaves us with a bit of breathing space when even longer options are introduced, as we don't have to reformat everything then.

File: outer.c [hunk 5]
```diff
@@ -279,36 +304,88 @@ struct outer_helper outer_helper_spawn(void)
...
-	int ok = 1;
-	write(pipefds_in[1], &ok, sizeof (ok));
+struct persistflag {
+	int flag;
+	const char *proc_ns_name;
+};
 
-	_exit(0);
+static struct persistflag pflags[] = {
+	{ BST_CLONE_NEWCGROUP, "cgroup" },
```
Comment (Snaipe) [inline outer.c:315]: Same comment as above about designated initializer syntax

File: outer.c [hunk 6]
```diff
@@ -279,36 +304,88 @@ struct outer_helper outer_helper_spawn(void)
...
+		if (nsfd < 0) {
+			if (errno != EEXIST) {
+				err(1, "creat(\"%s/%s\")", persist, f->proc_ns_name);
+			}
+		} else {
+			close(nsfd);
+		}
+		// Where is mountat()?  Thankfully, we can still name the persist directory.
+		char procname1[PATH_MAX];
+		snprintf(procname1, sizeof(procname1), "/proc/%d/ns/%s", pid, f->proc_ns_name);
```
Comment (Snaipe) [inline outer.c:348]: You should be able to just redefine procname here

File: bst.1.scd [hunk 7]
```diff
@@ -29,19 +29,37 @@ Users of bst may choose to opt-out of some of the isolation.
...
+	All namespaces are unshared by default; _e.g.,_ if _--share-mnt_
+	is not given, then the child process runs in a new (unshared)
+	mount namespace intialized as a copy of bst's parent's mount
+	namespace.
+
+\--persist <dir>
+	Bind-mount all namespaces of the new process into files in the
+	given directory.  The files are named the same as the
+	namespace files in /proc/[pid]/ns.  This option is equivalent
+	to bind-mounting /proc/[pid]/ns/* into <dir>/* after invoking
```
[the remark this answers is not in the document]
Comment (kduda) [reply bst.1.scd:58]: yep sorry.  fixed & proofread.

File: enter.c [hunk 8]
```diff
@@ -36,68 +37,59 @@
...
-		| BST_CLONE_NEWIPC
-		| BST_CLONE_NEWNS
-		| BST_CLONE_NEWNET
-		| BST_CLONE_NEWPID
-		| BST_CLONE_NEWUSER
-		| BST_CLONE_NEWUTS
-		| BST_CLONE_NEWTIME
-		,
+static struct cloneflag flags[] = {
+	{ BST_CLONE_NEWCGROUP, "cgroup" },
```
[the remark this answers is not in the document]
Comment (kduda) [reply enter.c:45]: done

File: enter.c (hunk 1 above)
Comment (kduda) [reply enter.c:65]: yep

File: enter.c (hunk 2 above)
Comment (kduda) [reply enter.c:127]: done, thanks, sorry for dropping this

File: enter.h (hunk 3 above)
Comment (kduda) [reply enter.h:19]: done

File: main.c (hunk 4 above)
Comment (kduda) [reply main.c:107]: an 8-char boundary, LOL.  I was wondering how you picked the column.

File: outer.c (hunk 5 above)
Comment (kduda) [reply outer.c:315]: done

File: outer.c (hunk 6 above)
Comment (kduda) [reply outer.c:348]: I actually just re-used the same procname local.

**Pull request merged**